underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AOutputStream.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.underfs.s3a;
+
+import alluxio.Configuration;
+import alluxio.Constants;
+import alluxio.util.io.PathUtils;
+
+import com.amazonaws.services.s3.internal.Mimetypes;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.PutObjectRequest;
+import com.amazonaws.services.s3.transfer.TransferManager;
+import com.amazonaws.util.Base64;
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.security.DigestOutputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A stream for writing a file into S3. The data will be persisted to a temporary directory on the
+ * local disk and copied as a complete file when the {@link #close()} method is called.
```

Comment:
Can you leave a comment for why we are not using the low level multipart streaming upload API? (http://docs.aws.amazon.com/AmazonS3/latest/dev/llJavaUploadFile.html). Since its an optimization which may not be required a TODO item may be sufficient.